### ISSUE TITLE
Custom component constructed multiple times regression

### ISSUE THREAD

**Issue by bpetex** [ngx-formly/ngx-formly#2285]
**Description**
After updating formly from 5.5.10 to 5.5.11 custom component is constructed two times instead of one. On the second construction prePopulate is not called. The view is used from the first construction. (Even if you for _buildForm after it will only call prePopulate inside the first component created not in the second). This seems like an undesired regression issue.

**Minimal Reproduction**
the issue with 5.5.11: https://stackblitz.com/edit/angular-xzoxnl?file=package.json
working version with 5.5.10: https://stackblitz.com/edit/angular-xzoxnl-5ya6d4?file=package.json

If you type in "whatever" to the input field in 5.5.10 it's validated correctly in 5.5.11 it shows a validation error even though it should be valid

**Comment by aitboudad** [comment]
Could you try the following solution https://stackblitz.com/edit/angular-xzoxnl-k7bybw

Well, that's currently a limitation where I didn't yet get a proper solution, the field prop should not have a hard dep of the component instance.

**Comment by bpetex** [comment]
That was just a very simple reproduction, my actual use case is very different. I know I can hack around it with passing the value to field.templateOPtions and it doesn't matter which component instance is checking, but this looks like a bug to me (a regression) which I didn't want to hack around. If you check the console logs in the first example, the second construction just doesn't make any sense at all, which did not happen in the previous version.

In case you think it's not a bug, please explain why it happens and why is it good/necessary

**Comment by aitboudad** [comment]
Well, sometimes introducing some features produce some regressions without noticing them until its too late :).

The issue exists since `v5` release but not noticeable and it occurs only when the field component is re-rendered and that the reason I decided to not fix this regression until finding a proper fix (a hidden bug is worse than the noticeable ones).

Anyway, I may give it a try again this week to see what I can do.

**Comment by bpetex** [comment]
So I was trying to update formly again, but it turned out to be quite impossible to hack around this one in complex custom components. @aitboudad can you please give an update when you can look into this one?

**Comment by aitboudad** [comment]
Not soon anyway the solution will be landed in the next major version and not `v5`.
In other side still not getting how my solution is complicated in your case (need the see the code in order to judge ^^).

if you want to get the field component instance whenever change you may try the following workaround:

```ts
export class TestTypeComponent extends FieldType {
  get field() { return this._field; }
  set field(f) {
    this._field = f;
    // field component changed
  }

  private _field;
}
```

hope this can help you a little!

**Comment by bpetex** [comment]
That doesn't help unfortunately, set only gets called in the first constructed instance not in the second. Can't really share the actual use case cause it's 4-5 related custom component with complex logic using form-arrays with the possibility multi layered groupings, changing and extending the recieved field definition on prePopulate and pleanty of logic in the component itself that I wouldn't consider adding to the field definition.

**Comment by aitboudad** [comment]
> Do you know which change caused in 5.5.11 this bug compared to 5.5.10, I might be able to just create a patch for it?

https://github.com/ngx-formly/ngx-formly/pull/2069

**Comment by intellix** [comment]
I think this is also the reason for this one: https://stackblitz.com/edit/ngx-formly-ui-material-bcsrs2?file=src%2Fapp%2Fcustom%2Fcustom.component.ts

We're stuck on 5.5.10 still and can't upgrade to fix another unrelated double-wrapping issue :)

**Comment by aitboudad** [comment]
@intellix that caused by using `this` inside `prePopulate`, `onPopulate` and `postPopulate`, use one of the following solutions instead:

1. Save `state` in `templateOptions` [custom.component.ts](https://stackblitz.com/edit/ngx-formly-ui-material-gcnfxm?file=src/app/custom/custom.component.ts)
2. Assign the component instance to the field [custom.component.ts](https://stackblitz.com/edit/ngx-formly-ui-material-mahit7?file=src/app/custom/custom.component.ts)  (for a workaround ⚠️)

**Comment by aitboudad** [comment]
I think providing the component instance through the field is the solution for this issue, it took a while to release it 😄, so I'll provide a fix where you can use `field.componentInstance` (any better name 🤔?) instead of `this`:

```patch
- this.state = 'b';
+ field.componentInstance.state = 'b';
```

**Note: this limitation only concern the calls inside the following functions: `prePopulate`, `onPopulate` and `postPopulate`.

**Comment by bpetex** [comment]
Sorry for the late reply. I'm updating atm for v6 next.
I was under the impression this was changed already, but it seems it was never implemented.
Yeah componentInstance sounds good.

**Comment by aitboudad** [comment]
This issue has been fixed and released as part of [v6.0.0-next.7](https://github.com/ngx-formly/ngx-formly/releases/tag/v6.0.0-next.7) release.

Please let us know, in case you are still encountering a similar issue/problem.
Thank you!

**Comment by bpetex** [comment]
Thanks :)
